### ISSUE TITLE
Assert in Builder::getContainedTypeId while compiling code with missing function definitions

### ISSUE THREAD

**Issue by zeux** [KhronosGroup/glslang#610]
When compiling code like this with entrypoint `main`:

```
float4 main(): SV_Target0 { return 0; }
struct Surface { float3 albedo; };
Surface surfaceShader(float fade);
Surface surfaceShaderExec()
{
    float fade = 0;
    return surfaceShader(0);
}
```

There is an assertion `assert(0)` that happens in Builder::getContainedTypeId with the following stack:

```
spv::Builder::getContainedTypeId(unsigned int typeId, int member) Line 582	C++
`anonymous namespace'::TGlslangToSpvTraverser::multiTypeStore(const glslang::TType & type, unsigned int rValue) Line 2431	C++
`anonymous namespace'::TGlslangToSpvTraverser::visitBranch(glslang::TVisit __formal, glslang::TIntermBranch * node) Line 1903	C++
glslang::TIntermBranch::traverse(glslang::TIntermTraverser * it) Line 264	C++
glslang::TIntermAggregate::traverse(glslang::TIntermTraverser * it) Line 170	C++
glslang::TIntermAggregate::traverse(glslang::TIntermTraverser * it) Line 170	C++
`anonymous namespace'::TGlslangToSpvTraverser::visitFunctions(const glslang::TVector<TIntermNode *> & glslFunctions) Line 2638	C++
`anonymous namespace'::TGlslangToSpvTraverser::visitAggregate(glslang::TVisit visit, glslang::TIntermAggregate * node) Line 1358	C++
glslang::TIntermAggregate::traverse(glslang::TIntermTraverser * it) Line 152	C++
glslang::GlslangToSpv(const glslang::TIntermediate & intermediate, std::vector<unsigned int,std::allocator<unsigned int> > & spirv, spv::SpvBuildLogger * logger) Line 5126	C++
glslang::GlslangToSpv(const glslang::TIntermediate & intermediate, std::vector<unsigned int,std::allocator<unsigned int> > & spirv) Line 5110	C++
```

What happens here is that during compiling the function call, the check in the first line of `TGlslangToSpvTraverser::handleUserFunctionCall` says that there is no function and it returns 0.

This causes the type checking during return processing (`TGlslangToSpvTraverser::visitBranch`, `EOpReturn` case) to consider the types as different because the IR essentially is malformed (the function call refers to the wrong instruction as a result of early-out in `handleUserFunctionCall`, triggering an assert.

First, it seems like there's some proper error handling that's missing here. I'm not sure if there is a valid case where handleUserFunctionCall should return 0 if the function isn't found - it does not seem like there is one?

Second, is there any way to resolve this so that the shader *does* get successfully converted to SPIR-V? I'm not very familiar with SPIR-V - in this case the correct thing to do would probably be to emit some sort of "extern" symbol for the `surfaceShader` function?

**Comment by zeux** [comment]
I guess these comments at the `handleUserFunctionCall` callsite:

```
        //assert(result);  // this can happen for bad shaders because the call graph completeness checking is not yet done
            logger->missingFunctionality("missing user function; linker needs to catch that");
```

Indicate that it's intentionally unsupported but error handling is currently lacking. Is it worth it uncommenting the assertion for now?

**Comment by johnkslang** [comment]
FYI, I'm considering instead of not bringing over dead subgraphs of the call graph. Have to finish checking on a few things before deciding (there are other DCE efforts).

**Comment by zeux** [comment]
Yeah - not bringing over dead subgraphs would also work. We are currently using spirv-remap from this repository to do DCE, but it might be helpful to just not generate the code in the first place.

**Comment by zeux** [comment]
FWIW this is now fixed for us (as of 906cc21) and our pre-spirv remap shader pack size is significantly reduced.
